server/src/main/java/de/uftos/repositories/ucdl/parser/DefinitionParser.java

```diff
@@ -0,0 +1,691 @@
+                    throw new ParseException(
+                        "ets can only be filtered using sets of the same type!");
+                  }
+                }
+                modifiers.add(new OperatorDto(UcdlToken.FILTER, filters));
+              }
+            }
+            parameters.replace("this", thisType);
+            return new SetDto(UcdlToken.RESOURCE_SET, ResourceType.NUMBER, numberSet, modifiers);
+          case "ELEMENT":
+            switch (root.jjtGetChild(1).toString()) {
+              case "SETMODIFICATION":
+                AbstractSyntaxTreeDto setName = buildAst(root.jjtGetChild(0), parameters);
+                ResourceType setType;
+                if (setName.getToken() == UcdlToken.NUMBER) {
+                  setType = ResourceType.NUMBER;
+                } else {
+                  setType = ((ElementDto) setName).type();
+                }
+
+                modifiers = new ArrayList<>();
+
+                modifierList = root.jjtGetChild(1);
+                while (modifierList.jjtGetNumChildren() > 0) {
+                  if (modifierList.jjtGetNumChildren() == 2) {
+                    SimpleNode flatmap = (SimpleNode) modifierList.jjtGetChild(0);
+                    String attribute = flatmap.jjtGetFirstToken().image;
+                    setType = getNextResourceType(setType, attribute);
+                    modifiers.add(new ValueDto<>(UcdlToken.ATTRIBUTE, attribute));
+                    modifierList = modifierList.jjtGetChild(1);
+
+                  } else if (modifierList.jjtGetNumChildren() == 3) {
+                    thisType = parameters.get("this");
+                    parameters.replace("this", setType);
+
+                    List<AbstractSyntaxTreeDto> filters = new ArrayList<>();
+                    filters.add(buildAst(modifierList.jjtGetChild(0), parameters));
+
+                    Node filterList = modifierList.jjtGetChild(1);
+                    while (filterList.jjtGetNumChildren() > 0) {
+                      filters.add(buildAst(filterList.jjtGetChild(0), parameters));
+                      filterList = filterList.jjtGetChild(1);
+                    }
+
+                    for (AbstractSyntaxTreeDto filter : filters) {
+                      if ((filter.getToken() == UcdlToken.NUMBER_SET
+                          && setType != ResourceType.NUMBER)
+                          || (filter.getToken() == UcdlToken.RESOURCE_SET
+                          && ((SetDto) filter).type() != setType)) {
+                        throw new ParseException(
+                            "Sets can only be filtered using sets of the same type!");
+                      }
+                    }
+
+                    modifiers.add(new OperatorDto(UcdlToken.FILTER, filters));
+                    modifierList = modifierList.jjtGetChild(2);
+                    parameters.replace("this", thisType);
+                  }
+                }
```

Comment:
I *think* I've seen that code block already, but it's also just very complex

Reply:
Probably also exists in `case "SET"`...
I need to refactor a lot of code duplication.